hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/stack/savepoints/AbstractSavepoint.java

```diff
@@ -0,0 +1,153 @@
+            @NonNull final WrappedHederaState state,
+            @NonNull final BuilderSink parentSink,
+            final int maxPreceding,
+            final int maxFollowing) {
+        super(maxPreceding, maxFollowing);
+        this.state = requireNonNull(state);
+        this.parentSink = requireNonNull(parentSink);
+    }
+
+    /**
+     * Constructs a savepoint with a total limit on the number of builders that can be accumulated.
+     * @param state the current state
+     * @param parentSink the parent sink
+     * @param maxTotal the maximum number of total builders
+     */
+    protected AbstractSavepoint(
+            @NonNull final WrappedHederaState state, @NonNull final BuilderSink parentSink, final int maxTotal) {
+        super(maxTotal);
+        this.state = requireNonNull(state);
+        this.parentSink = requireNonNull(parentSink);
+    }
+
+    @Override
+    public HederaState state() {
+        return state;
+    }
+
+    @Override
+    public void commit() {
+        state.commit();
+        commitBuilders();
+    }
+
+    @Override
+    public void rollback() {
+        rollback(precedingBuilders);
+        rollback(followingBuilders);
+        commitBuilders();
+    }
+
+    @Override
+    public Savepoint createFollowingSavepoint() {
+        return new FollowingSavepoint(new WrappedHederaState(state), this);
+    }
+
+    @Override
+    public SingleTransactionRecordBuilder createBuilder(
+            @NonNull final SingleTransactionRecordBuilder.ReversingBehavior reversingBehavior,
+            @NonNull final HandleContext.TransactionCategory txnCategory,
+            @NonNull final ExternalizedRecordCustomizer customizer,
+            final boolean isBaseBuilder) {
+        requireNonNull(reversingBehavior);
+        requireNonNull(txnCategory);
+        requireNonNull(customizer);
+        final var builder = new SingleTransactionRecordBuilderImpl(reversingBehavior, customizer, txnCategory);
+        if (!customizer.shouldSuppressRecord()) {
+            if (txnCategory == PRECEDING && !isBaseBuilder) {
+                addPrecedingOrThrow(builder);
+            } else {
+                addFollowingOrThrow(builder);
+            }
+        }
+        return builder;
+    }
+
+    /**
+     * Commits the builders accumulated in this savepoint to the parent sink.
+     */
+    abstract void commitBuilders();
+
+    private void rollback(@NonNull final List<SingleTransactionRecordBuilder> builders) {
+        var removedBuilder = false;
+        for (int i = 0; i < builders.size(); i++) {
+            final var builder = builders.get(i);
+            if (builder.reversingBehavior() == REVERSIBLE) {
+                builder.nullOutSideEffectFields();
+                if (SUCCESSES.contains(builder.status())) {
+                    builder.status(REVERTED_SUCCESS);
+                }
+            } else if (builder.reversingBehavior() == REMOVABLE) {
+                builders.set(i, null);
```

Comment:
I think you should use a `ListIterator` instead.

Reply:
Yes fixed. Thank you!